### ISSUE TITLE
Compatibility with Unity IAM

### ISSUE THREAD

**Issue by olifre** [scitokens/scitokens-cpp#53]
Not sure if this is out of scope for `scitokens-cpp`, but I'll collect my experiences here and leave the decision to you ;-). 

I've been trying to get the scitokens-cpp library working with the Unity IAM, used e.g. by the Helmholtz AAI ( https://login.helmholtz.de/oauth2 ) and B2ACCESS ( https://b2access.eudat.eu/home/ ) . 

I've hit two issues:
* It bails out at https://github.com/scitokens/scitokens-cpp/blob/7d9c77bd0fd238a8db000acd8229616e786ba273/src/scitokens_internal.h#L264-L266 since the Unity IAM does not put the `nbf` claim inside: https://github.com/unity-idm/unity/blob/a07e075c0287584e56705c416c3f0280c23efafd/rest/src/main/java/pl/edu/icm/unity/rest/jwt/JWTUtils.java#L34-L38 . According to [rfc7515](https://datatracker.ietf.org/doc/html/rfc7515#section-4.1.4), it seems `nbf` is regarded as optional (they provide `iat` which is validated by `jwt-cpp` ). `jwt-cpp` regards `nbf` as optional. 
Since there are no explicit requirements in the code apart from the linked check, I commented these lines out for my test. 
* I've run into the issue addressed in https://github.com/scitokens/scitokens-cpp/pull/7 . Trying with the patch from that issue, `find_key_id` failed: https://github.com/scitokens/scitokens-cpp/blob/7d9c77bd0fd238a8db000acd8229616e786ba273/src/scitokens_internal.cpp#L436 since it tried to find the key by `kid` from the metadata, which of course fails with an empty `kid`. 

I think these are all issues, but I stopped at this point to ask the question: Is there interest to get compatibility with the Unity IAM, or is this regarded as out of scope for this library? 
If there's interest in compatibility, I could try to continue. My proposal then would be to make `nbf` optional, and revive the linked PR to make `find_key_id` work in case `kid` is missing.

**Comment by djw8605** [comment]
Yes, I think there is a desire to be as compatible as possible, and you certainly pointed out bugs.  We are more than happy to receive pull requests for these changes, or if you would like help, we could help craft the changes.

**Comment by olifre** [comment]
Thanks a lot for this positive judgement! Then I'll continue on this work (potentially next week if time allows) and come up with PRs. Some things will be easy to change, but I may come back and ask about the key-ID handling — but I'll read up on the existing code and the RFCs first.

**Comment by djw8605** [comment]
I was reading through the [SciTokens verification rules](https://scitokens.org/technical_docs/Claims).  They say that `nbf` is required.  The [WLCG Profile](https://zenodo.org/record/3460258#.YOcBxxNudf0) has `nbf` as optional.  We could always add a new profile, if you are not using the WLCG profile, to the library.

**Comment by bbockelm** [comment]
I do think it would be a good idea to add a "plain JWT" profile that _only_ does what is in the RFC and doesn't have any of the processing rules for SciTokens / WLCG.

@olifre - would it be possible to show the public portion of the token that Unity IAM generates?

**Comment by olifre** [comment]
@djw8605 Indeed, an additional profile (and that might be "plain JWT" and / or something for the Unity IAM) sounds like a good plan. 

@bbockelm You mean the header and payload of a fetched token, correct? 
Here's the header:
```json
{
  "typ": "at+jwt",
  "alg": "RS256"
}
```
and here's the payload:
```json
{
  "sub": "<uuid>",
  "aud": "public-oidc-agent",
  "scope": "openid profile",
  "iss": "https://login.helmholtz.de/oauth2",
  "exp": 1625756628,
  "iat": 1625752628,
  "jti": "<uuid>",
  "client_id": "public-oidc-agent"
}
```

With this, the current issues are:
* `nbf` missing
* `kid` not present. 

So in principle, a Unity IAM profile (removing these two requirements) could be used, but a "plain JWT" would for sure also be interesting.

**Comment by bbockelm** [comment]
@djw8605 - per (the new!) [RFC8725 Section 3.11](https://datatracker.ietf.org/doc/html/rfc8725#section-3.11), the right way to handle the type of the token is through the `typ` parameter in the header.  It also recommends that one makes these unique for each type (or what we're calling profile).

It looks like the `at+jwt` type is being standardized here:
https://datatracker.ietf.org/doc/html/draft-ietf-oauth-access-token-jwt-13

So, this profile could be (a) be auto-detected based on `typ="at+jwt"` and (b) apply the rules specified in the draft "JSON Web Token (JWT) Profile for OAuth 2.0 Access Tokens" document.

I think we should open a separate ticket for "kid not present in the header" -- I'm not sure that's really related to the profile.  I think it's bad form for Unity to not identify the key ID used but I am OK with the idea of defaulting to a key if there's only a single key present.

**Comment by olifre** [comment]
> I think we should open a separate ticket for "kid not present in the header" -- I'm not sure that's really related to the profile. I think it's bad form for Unity to not identify the key ID used but I am OK with the idea of defaulting to a key if there's only a single key present.

That sounds good, we could also take that part here which already has parts of the necessary changes: https://github.com/scitokens/scitokens-cpp/pull/7

I can also try to raise this as an issue / feature request to the Unity developers in parallel (their bug reporting seems to work via their [discussion mailing list](https://www.unity-idm.eu/support/), since they have GitHub issues disabled and only offer a [read-only public bug tracker](https://unity-idm.atlassian.net/jira/software/c/projects/UY/issues/), which is quite strange for an OpenSource project). So I'm not too optimistic on getting that end to behave better, but I'll give it a try.

**Comment by olifre** [comment]
@bbockelm @djw8605 A happy new year to both of you! 

PR #73 now addresses the remaining part, after applying #55 and #73 I can successfully authenticate to e.g. XRootD using the scitokens library with a token from the Helmholtz AAI. 

Comments and improvements to the PRs are very appreciated :smile: .

**Comment by olifre** [comment]
@djw8605 Thanks for reviewing both PRs and merging them so quickly :smile:. 

I'll also close this issue here then, since tokens produced by the Unity IAM now work fine. The tokens produced by the Unity IAM currently do not seem to contain group information, but that's something which has to be solved and later handled elsewhere.

**Comment by djw8605** [comment]
@olifre do you need an EPEL release quickly?  I wasn't planning on another EPEL release for a bit, but could if it's a priority.

**Comment by olifre** [comment]
@djw8605 Thanks for asking! 
It depends a bit on what you mean with "quickly" — it would be nice / ideal to be able to use this in production some time this month, since I also know some other parties are interested in this feature (the Unity IAM is used by some Research Data projects in Europe, while they have not yet settled on it for the final operation, it's the current choice for pilot projects).
 They would of course also be happy about an EPEL release to ease deployment in production. 

So a release some time this month would indeed be nice, if possible and not too much work :smile: .

**Comment by timtheisen** [comment]
Last time I tried to build on Debian, recent changes caused it to not build.
I'll try again and see if the problem persists.

...Tim

--
Tim Theisen (he, him, his)
Release Manager
HTCondor & Open Science Grid
Center for High Throughput Computing
Department of Computer Sciences
University of Wisconsin - Madison
4261 Computer Sciences and Statistics
1210 W Dayton St
Madison, WI 53706-1685
+1 608 265 5736
________________________________

**Comment by olifre** [comment]
@djw8605 Just a kind reminder after a month has passed ;-). Getting the helper binaries within a release as pointed out by @DrDaveD also seems very useful. 

@timtheisen Compilation on Debian 11 works fine for me with the current master branch.

**Comment by djw8605** [comment]
Released.  Just need some karma in EPEL:
- EL8 https://bodhi.fedoraproject.org/updates/FEDORA-EPEL-2022-d65014977c
- EL9 https://bodhi.fedoraproject.org/updates/FEDORA-EPEL-2022-76276422c0
- F34 https://bodhi.fedoraproject.org/updates/FEDORA-2022-ca2909a0c8
- EL7 https://bodhi.fedoraproject.org/updates/FEDORA-EPEL-2022-7bbf95c6f2
- F35 https://bodhi.fedoraproject.org/updates/FEDORA-2022-0c21c20f76
